lib/spack/spack/spec.py

```diff
@@ -2221,24 +2234,23 @@ def normalize(self, force=False, tests=False):
         # Get all the dependencies into one DependencyMap
         spec_deps = self.flat_dependencies(copy=False)
 
+        full_spec_deps = dict(spec_deps)
+        if user_spec_deps:
+            for name, spec in user_spec_deps.items():
+                if name not in full_spec_deps:
```

Comment:
Won't this skip user constraints whose name matches a package constraint?  I think you want to add:

```python
    ...
else:
    full_spec_deps[name].constrain(spec)
```